packages/babel-core/src/config/option-manager.js

```diff
@@ -70,23 +71,84 @@ const ALLOWED_PLUGIN_KEYS = new Set([
   "post",
   "visitor",
   "inherits",
+  "capabilities",
+  "dependencies",
 ]);
 
 export default function manageOptions(opts: {}): {
   options: Object,
-  passes: Array<Array<[ Plugin, ?{} ]>>,
+  plugins: Array<[ Plugin, ?{} ]>,
 }|null {
   return new OptionManager().init(opts);
 }
 
+function sortPlugins(plugins) {
+  const graph = new DagMap();
+  const capabilitiesToPluginIdMap = Object.create(null);
+
+  // It is theoretically possible that something arrives at this point without
+  // having a name or key property. Make sure that it has one.
+  const unnamedPluginPrefix = "unnamedPlugin";
+  plugins.forEach((pluginTuple, index) => {
+    const plugin = pluginTuple[0];
+    if (!plugin.key) { plugin.key = `${unnamedPluginPrefix}${index}`; }
+  });
+
+  // Calculate the capabilities that each plugin provides, map to a single name.
+  // Multiple plugins with the same capability are not allowed as that can
+  // become a cyclic dependency.
```

Comment:
I don't think it will become a cyclic dependency - don't think their are any actual ones? Our primary goal is to fix plugin ordering in Babel itself (only syntax), then minifier, then 3rd party plugins (not as important).

It's just about telling the user if a plugin is included more than once just because it doesn't have to be.